### PR TITLE
Rappel system for tadpole

### PR DESCRIPTION
**Pull request by breadhunt**
## About The Pull Request
A revival of https://github.com/tgstation/TerraGov-Marine-Corps/pull/11636 - see that PR for mechanics

**PLEASE TM THIS PR** - I have tested it but I'm not 100% sure (+ git fuckery has probably messed something up)

This PR should be completely fine to go on its own, but I'm planning a few extra tadpole changes to shift its role as well. See https://hackmd.io/zJlQF-T1QNSvOA1APFvrNw.

Also adds vendors near tadpole for convenience, probably should be atomised but w/e

## Why It's Good For The Game

There are a few reasons for this PR:
1. It's cool.
2. Provides a tool for extra tactical mobility. How is this any different from pods? Pods are shipside, which if you're going alone is great, but it's an ass to transport 6 marines from groundside --> shipside --> podding down. It _should_ be a lot easier to do the same thing, but with Tadpole instead!
3. Carves a new niche for Tadpole. Tadpole as a mini-fortress is boring. Right now it has no clear role and it's frankly frustrating to play... _unless_ you're playing it as a transport. Transport tadpole is very very cool, fits what it was originally envisioned as (a dropship) and should allow for some cool teamplay. (We'll see)

TL;DR pair with Researcher, pair with a squad on a flank, rescue the dead, etc. Go have fun.
I have some potential ideas for nerfs/buffs in mind if it's over/undertuned but feel free to raise any walance concerns you have about it
## Changelog
:cl: tyeagg for original code, breadhunt for reviving it
add: Adds rappel system, turning POs into glorified elevators once more.
add: Adds the spare rappel cord box. Required to repair the rappel system.
imageadd: rappel system sprite, rappel cord box
/:cl:

### CODE REVIEW

**Comment by Lumipharon** [comment]
I have some balance doubts, but happy to try it out

**Comment by Gregormack** [comment]
I'm mostly sure it is a complication to the non existing problem. But, whatever

**Comment by SimplySmiley123** [comment]
Would the Tadpole still be able to land groundside to:

a) deliver supplies like ammo, welding kits, etc.

b) act as a beacon for requisitions

c) be usable by a researcher to process samples

d) transport the vehicles from shipside to groundside, and on groundside from location A to location B

e) be usable as evac for nuke

**Comment by lbnesquik** [comment]
You know what? I think this is cool. I think the cool factor is undeniable and that's something we shouldn't forget in this game. Also, please make it so xenos can go up the ropes.

**Comment by Lumipharon** [comment]
@breadhunt do you want to just put the rappel in its own file? the dropship equipment file is already a mess of a gorillion different things that need splitting up.

File: code/game/objects/structures/dropship_equipment.dm [hunk 1]
```diff
@@ -877,42 +927,356 @@
...
+/obj/structure/dropship_equipment/rappel_system/examine(mob/user)
+	. = ..()
+	switch(rappel_condition)
+		if(RAPPEL_CONDITION_DISABLED)
+			. += "It's smoking. It'll take some time for its self-repair function to kick in."
+		if(RAPPEL_CONDITION_DAMAGED)
+			. += "The rappel's cord is completely broken and needs replacing."
+
+
+///Human interaction with the rappel system; this is how people rappel down
```
Comment (Lumipharon) [inline code/game/objects/structures/dropship_equipment.dm:1017]: don't autodoc preexisting procs

File: code/game/objects/structures/dropship_equipment.dm [hunk 2]
```diff
@@ -877,42 +927,356 @@
...
+
+	switch(rappel_condition)
+		if(RAPPEL_CONDITION_DAMAGED)
+			balloon_alert(user, "The cord needs replacing!")
+			return
+		if(RAPPEL_CONDITION_DISABLED)
+			balloon_alert(user, "The system is disabled!")
+			return
+
+	if(!is_reserved_level(z))
```
Comment (Lumipharon) [inline code/game/objects/structures/dropship_equipment.dm:1030]: does this make it work in space, as opposed to just cloud land?

File: code/game/objects/structures/dropship_equipment.dm [hunk 3]
```diff
@@ -877,42 +927,356 @@
...
+			balloon_alert(user, "The rappel is currently retracting!")
+			return
+
+	var/turf/target_turf = get_turf(rope)
+	if(target_turf.density)
+		balloon_alert(user, "You can't rappel into a wall!")
+		return
+
+	var/area/target_area = get_area(target_turf)
+	if(target_area.ceiling >= CEILING_DEEP_UNDERGROUND)
```
Comment (Lumipharon) [inline code/game/objects/structures/dropship_equipment.dm:1048]: this lets you rappel into caves and in buildings?

File: code/game/objects/structures/dropship_equipment.dm [hunk 4]
```diff
@@ -877,42 +927,356 @@
...
+	flick("rope_up", rope)
+
+	addtimer(CALLBACK(src, .proc/reel_in), 0.4 SECONDS)
+
+///Part 2 of retract_rope(), moves the rope back into the system after the rope animation has completed. Unbuckles any mobs which were attached to it.
+/obj/structure/dropship_equipment/rappel_system/proc/reel_in()
+	rope.forceMove(src)
+
+///Handles xeno attacks on the system; called by the rappel rope whenever attack_alien() is called on the rope
+/obj/structure/dropship_equipment/rappel_system/proc/buckle(mob/living/carbon/xenomorph/X)
```
Comment (Lumipharon) [inline code/game/objects/structures/dropship_equipment.dm:1181]: don't use X.
Also this probably should have another name since its just confusing it with actual buckling.

File: code/game/objects/structures/dropship_equipment.dm [hunk 5]
```diff
@@ -877,42 +927,356 @@
...
+
+	rappel_condition = RAPPEL_CONDITION_DAMAGED
+	rappel_state = RAPPEL_STATE_RETRACTING
+	update_icon_state()
+	retract_rope()
+
+///Ghosts teleport to the rappel system when they click on the rope
+/obj/effect/rappel_rope/tadpole/attack_ghost(mob/dead/observer/user)
+	. = ..()
+	user.forceMove(get_turf(parent_system))
```
Comment (Lumipharon) [inline code/game/objects/structures/dropship_equipment.dm:1224]: put this under the obj definition

File: code/game/objects/structures/dropship_equipment.dm [hunk 6]
```diff
@@ -877,42 +927,356 @@
...
+
+///Going up the rappel. Going up retracts the rappel
+/obj/effect/rappel_rope/tadpole/attack_hand(mob/living/user)
+	. = ..()
+
+	if(parent_system.rappel_condition < RAPPEL_CONDITION_GOOD)
+		return
+
+	if(user.buckled_mobs)
+		//Drop anyone we're fireman carrying
```
Comment (Lumipharon) [inline code/game/objects/structures/dropship_equipment.dm:1257]: sad

File: code/game/objects/structures/dropship_equipment.dm [hunk 7]
```diff
@@ -877,42 +927,356 @@
...
+
+	addtimer(CALLBACK(src, .proc/reel_in), 0.4 SECONDS)
+
+///Part 2 of retract_rope(), moves the rope back into the system after the rope animation has completed. Unbuckles any mobs which were attached to it.
+/obj/structure/dropship_equipment/rappel_system/proc/reel_in()
+	rope.forceMove(src)
+
+///Handles xeno attacks on the system; called by the rappel rope whenever attack_alien() is called on the rope
+/obj/structure/dropship_equipment/rappel_system/proc/buckle(mob/living/carbon/xenomorph/X)
+	//Stops the pilot bringing up the rappel to prevent it being disabled
```
Comment (Lumipharon) [inline code/game/objects/structures/dropship_equipment.dm:1182]: what happens if tad just flies off/land while a xeno is breaking things? Or does that stop it somewhere? I don't see it.

**Review by Lumipharon — CHANGES_REQUESTED** [review]

**Comment by iforgotmeword** [comment]
"Tadpole as a mini-fortress is boring" 
it's not however it was busted as all hell 

i would love to see it return be it much much weaker

overall i think this is a good change 
(also Tad could be the new way req transports stuff)

File: code/game/objects/structures/dropship_equipment.dm (hunk 6 above)
Comment (breadhunt) [reply code/game/objects/structures/dropship_equipment.dm:1257]: Tadpole should probably have to have land clear in order to rescue bodies

**Comment by breadhunt** [comment]
> I feel like, with this new spooling mechanic, the marines will need more ways to stop the xenos from just camping around a corner and shitting on their engines... maybe better TAD turret placement? otherwise its just having the marines leave the some-what safety of the TAD to punch a xeno a few times, so the TAD can leave, risking the marine getting murdered and/or left behind because the entire T4 force of the xenos is munching on fiber optic cables
> 
> edit: i would also say that you should either make ropes craftable via the TAD/CAS fabricator, or make it so you can repair the actual rappel system yourself, because in-game 10 minutes is a REALLY long time, and is a big amount of the round, when you think about it.
> 
> heres an example of this: if you play helldivers 2, and you have used the mech before, you would know that you only get 2 mechs per mission, if lucky, and that they have like a... 10-20 minute cooldown to spawn, last i checked. that means, even if you had more than 2 mechs, you would only get to spawn 2 max, due to the longest missions only being 40 minutes long
> 
> in TGMC, we tend to have missions that last from 20 minutes (shitrines or fenoflop) to maybe even 2 hours, if (un)lucky. this means, with 10 minutes to repair, you would get, most likely, 2-5 uses per 20 minute game, due to marines still being competent enough that they will defend the rope until you can retract it.
> 
> my point being... probably shorten that to 5 minutes to repair the internal mechanism?
> 
> also, i aint readin allat (i predict)

I aint readin allat

It was originally set to 5 minutes but I think it needs to be punishing to use, considering how powerful it is, but I'll tweak it depending on how it goes

Outside the scope of this PR but...
The intention with spooling is for tadpole to be better as a miscellaneous transport vehicle and worse as a combat vehicle. If marines want to secure a tadpole LZ for launching/landing they _should_ have to cade it, put a minefield around it, whatever. As I said in the original PR there's a reason maints want to get rid of tadpole altogether.

**Comment by Lumipharon** [comment]
Don't self resolve review comments please

File: code/game/objects/structures/dropship_equipment.dm (hunk 7 above)
Comment (Lumipharon) [reply code/game/objects/structures/dropship_equipment.dm]: you didn't answer this

Comment (breadhunt) [reply code/game/objects/structures/dropship_equipment.dm]: I've added an interaction for it, it just breaks the rappel & the tad flies off

**Review by Lumipharon — APPROVED** [review]

**Comment by SiegeB0t** [comment]
Looks like the rope is getting qdeleted somewhere breaking the console. Maybe since its an effect or maybe from xeno acid. Its deletion should probably be handled as if it were destroyed by some unknown weapon.

Example:
1. Recently did rappel op where FC set the rappel and landed a few marines.
2. Tad landed and xenos invaded.
3. Tad took off.
4. TO tried to use the rappel but got balloons "You shouldn't be seeing this; ask admins for help!" and "The rappel is currently retracting!"

**Comment by TiviPlus** [comment]
checks failing due to runtime btw

**Comment by DrAmazing343** [comment]
Played the greatest round of my life when this was TM'd- Worked with Charlie squad to perform some insane fucking digsite smash and grabs for samples, drawing heat from the hive after rappelling and clearing LZ, then hot extracting right as the first of the T4's pulled up. Zig-zagged into cave drops and back up, pulling out whenever the heat got too hot. It was fucking fantastic. This is all I could ever ask for for the TAD in today's TGMC.

**Comment by Breadeagle** [comment]
Rappels should probably have a skill lock if they don't already. Mostly for researchers. There's some pretty degenerate things you can do with a movable, hidden, no landing required rappel and researcher. Especially when digsites are hidden to xenos. 

Visible digsites help but they lead to more of the same, tedious gameplay that counters researchers right now, which is gargoyle spam. Both should probably happen either way I guess.

**Comment by DrAmazing343** [comment]
>skill lock
> 
> gargoyle spam

I've begun seeing a lot more gargoyle spam in the past couple of days after playing a shitload of RSR; Skill LOCK would hurt quite a bit, but I'd be 100% down for a skill-slow on this. It makes sense a Xenobiologist would fumble on the line a bit, and encourage marines to go first and fan out, giving more warning time for the gargoyles as LZ is cleared.

I don't think the dynamic between the two is necessarily bad, but a lock would put a helluva damper on my feelings about this PR.

**Comment by SimplySmiley123** [comment]
> Rappels should probably have a skill lock if they don't already. Mostly for researchers. There's some pretty degenerate things you can do with a movable, hidden, no landing required rappel and researcher. Especially when digsites are hidden to xenos.
> 
> Visible digsites help but they lead to more of the same, tedious gameplay that counters researchers right now, which is gargoyle spam. Both should probably happen either way I guess.

There's gonna be gargoyle spam regardless of wether RSRs use tadpole at all or not, since gargoyles ain't even solely because of researchers but also because of miners and guaranteeing no one secures shit outside of the main ball of marines.

**Comment by Lumipharon** [comment]
Probs needs a faster channel time to rappel DOWN (or just make it auto rappel if you walk on top).
Also fix conflicts nerd.

**Comment by breadhunt** [comment]
Rappel-down time reduced to 2s. 3 second fumble time added whenever civilians rappel up/down.

File: code/game/objects/structures/rappel_system.dm [hunk 8]
```diff
@@ -132,6 +132,11 @@
 	user.client.perspective = EYE_PERSPECTIVE
 	user.client.eye = target_turf
 
+	if(user.skills.getRating(SKILL_FIREARMS) < SKILL_FIREARMS_DEFAULT)
```
Comment (Lumipharon) [inline code/game/objects/structures/rappel_system.dm:135]: this seems pretty smelly relating the firearm skill with an action entirely unrelated to firearms. I won't stop it from being tested, but I don't like this at all.

**Review by Lumipharon — CHANGES_REQUESTED** [review]

**Comment by Lumipharon** [comment]
nevermind, conflicts anyway

**Comment by Lumipharon** [comment]
Not forgotten, just conflciting with other TM's.

**Comment by Lumipharon** [comment]
@breadhunt resolve conflict

**Comment by breadhunt** [comment]
Fuck this I'm figuring it out tomorrow

**Comment by breadhunt** [comment]
Continuing on https://github.com/tgstation/TerraGov-Marine-Corps/pull/16282
